File: src/main/java/de/moritzf/opencodewebpanel/toolWindow/OpenCodeWebToolWindowFactoryImpl.java
package de.moritzf.opencodewebpanel.toolWindow;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.wm.ToolWindow;
import com.intellij.openapi.wm.ToolWindowFactory;
import com.intellij.ui.content.Content;
import com.intellij.ui.content.ContentFactory;
import org.jetbrains.annotations.NotNull;

public final class OpenCodeWebToolWindowFactoryImpl implements ToolWindowFactory, DumbAware {
    @Override
    public void createToolWindowContent(@NotNull Project project, @NotNull ToolWindow toolWindow) {
        OpenCodeWebToolWindowFactory.OpenCodeWebToolWindowContent toolWindowContent =
            new OpenCodeWebToolWindowFactory.OpenCodeWebToolWindowContent(toolWindow);
        ApplicationManager.getApplication().invokeLater(() -> {
            Content content = ContentFactory.getInstance().createContent(toolWindowContent.getContent(), null, false);
            content.setDisposer(toolWindowContent);
            toolWindow.getContentManager().addContent(content);
            toolWindowContent.checkAndLoadContent();
        });
    }
}

File: src/main/kotlin/de/moritzf/opencodewebpanel/toolWindow/OpenCodeWebToolWindowFactoryImpl.kt
package de.moritzf.opencodewebpanel.toolWindow

import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.project.Project
import com.intellij.openapi.wm.ToolWindow
import com.intellij.openapi.wm.ToolWindowFactory
import de.moritzf.opencodewebpanel.server.OpenCodeServerBackend
import de.moritzf.opencodewebpanel.server.OpenCodeServerBackendRegistry
import com.intellij.ui.content.ContentFactory
import kotlin.jvm.JvmDefaultWithoutCompatibility

internal fun openCodeToolWindowHeading(project: Project?): String {
    val native = OpenCodeServerBackend.isNative(
        OpenCodeServerBackendRegistry.getInstance().backendFor(project).backendId,
    )
    return if (native) "OpenCode (native CLI)" else "OpenCode (sbx)"
}

internal fun updateOpenCodeToolWindowHeading(toolWindow: ToolWindow) {
    if (toolWindow.isDisposed) return
    toolWindow.stripeTitle = openCodeToolWindowHeading(toolWindow.project)
}

@JvmDefaultWithoutCompatibility
class OpenCodeWebToolWindowFactoryImpl : ToolWindowFactory, DumbAware {
    override fun createToolWindowContent(project: Project, toolWindow: ToolWindow) {
        val controller = OpenCodePanelController.getInstance(project)
        val content = ContentFactory.getInstance().createContent(controller.toolWindowComponent, null, false)
        toolWindow.contentManager.addContent(content)
        controller.attachToolWindow(toolWindow)
        updateOpenCodeToolWindowHeading(toolWindow)
        installTitleActions(toolWindow)
    }

    private fun installTitleActions(toolWindow: ToolWindow) {
        // Icon-only actions in the existing title bar; IntelliJ clips them on narrow panels,
        // so the gear menu below duplicates everything.
        toolWindow.setTitleActions(openCodeTitleActions())
        toolWindow.setAdditionalGearActions(openCodeGearActions())
    }
}
